### ISSUE TITLE
`torch.compile` silent bug in GGUF dequant, possibly because of `view(dtype)`

### ISSUE THREAD

**Issue by woct0rdho** [pytorch/pytorch#172747]
### 🐛 Describe the bug

I caught this bug when implementing GGUF dequant in torch. I've started related discussions at https://github.com/city96/ComfyUI-GGUF/issues/405 and https://github.com/huggingface/transformers/issues/40070 . I've tried my best to make a minimal reproducer:
```python
import torch

def generate_blocks_bf16(n_blocks):
    data = torch.zeros((n_blocks, 2), dtype=torch.uint8)
    data[:, 0] = 0x80
    data[:, 1] = 0x3F
    return data

def generate_blocks_Q6_K(n_blocks):
    data = torch.zeros((n_blocks, 210), dtype=torch.uint8)
    data[:, 192:208] = 1
    data[:, 208] = 0x00
    data[:, 209] = 0x3C
    return data

def dequantize_blocks_BF16(blocks, dtype):
    return (blocks.view(torch.int16).to(torch.int32) << 16).view(torch.float32)

def split_block_dims(blocks, *args):
    n_max = blocks.shape[1]
    dims = list(args) + [n_max - sum(args)]
    xs = torch.split(blocks, dims, dim=1)
    return xs

def view_float16(x):
    return x.view(torch.float16)

def dequantize_blocks_Q6_K(blocks, dtype):
    QK_K = 256
    (ql, qh, scales, d) = split_block_dims(blocks, QK_K // 2, QK_K // 4, QK_K // 16)

    scales = scales.view(torch.int8).to(dtype)
    d = view_float16(d).to(dtype)
    d = d * scales

    n_blocks = blocks.shape[0]
    ql = ql.reshape((n_blocks, -1, 1, 64))

    return d

def wrap_compile(func):
    @torch.compile
    def _func(blocks, dtype):
        return func(blocks, dtype)

    return _func

dequantize_blocks_BF16_compiled = wrap_compile(dequantize_blocks_BF16)
dequantize_blocks_Q6_K_compiled = wrap_compile(dequantize_blocks_Q6_K)

device = "cpu"
n_blocks = 32

blocks_bf16 = generate_blocks_bf16(n_blocks).to(device)
blocks_Q6_K = generate_blocks_Q6_K(n_blocks).to(device)

out_bf16 = dequantize_blocks_BF16(blocks_bf16, torch.float32)
print("BF16 mean", out_bf16.mean().item())

out_Q6_K = dequantize_blocks_Q6_K(blocks_Q6_K, torch.float32)
print("Q6_K mean", out_Q6_K.mean().item())

out_bf16_compiled = dequantize_blocks_BF16_compiled(blocks_bf16, torch.float32)
print("BF16 mean compiled", out_bf16_compiled.mean().item())

out_Q6_K_compiled = dequantize_blocks_Q6_K_compiled(blocks_Q6_K, torch.float32)
print("Q6_K mean compiled", out_Q6_K_compiled.mean().item())
```
To pass the test, "Q6_K mean" should be the same as "Q6_K mean compiled". When the test fails, the output is like
```
BF16 mean 1.0
Q6_K mean 1.0
BF16 mean compiled 1.0
Q6_K mean compiled 0.0625
``` 
It fails on both CPU and CUDA.

Interestingly, if I comment out `out_bf16_compiled = dequantize_blocks_BF16_compiled(blocks_bf16, torch.float32)`, then the test passes.

Although `ql = ql.reshape((n_blocks, -1, 1, 64))` does not affect `d`, if I comment out it, the test passes.

I guess there is some undefined behavior like out-of-bounds memory access when we call `view(dtype)` and the number of bytes of the element changes.

Log with `TORCH_TRACE`: [dedicated_log_torch_trace_gox8spt4.log](https://github.com/user-attachments/files/24696736/dedicated_log_torch_trace_gox8spt4.log)

### Error logs

No error. It silently gives wrong numbers or nan.

### Versions

PC with Nvidia GPU, Windows, torch 2.10 (RC version):
<details>
<summary>Expand</summary>

Collecting environment information...
PyTorch version: 2.10.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Enterprise (10.0.26200 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 4.2.1
Libc version: N/A

Python version: 3.12.10 (tags/v3.12.10:0cc8128, Apr  8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Ti Laptop GPU
Nvidia driver version: 591.59
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Name: 12th Gen Intel(R) Core(TM) i9-12900H
Manufacturer: GenuineIntel
Family: 207
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 2500
MaxClockSpeed: 2500
L2CacheSize: 11776
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] clip-anytorch==2.6.0
[pip3] cublas_ops==0.0.5
[pip3] dctorch==0.1.2
[pip3] flake8==7.3.0
[pip3] flake8-2020==1.8.1
[pip3] flake8-bandit==4.1.1
[pip3] flake8-broken-line==1.0.0
[pip3] flake8-bugbear==25.11.29
[pip3] flake8-builtins==3.1.0
[pip3] flake8-comprehensions==3.17.0
[pip3] flake8-datetimez==20.10.0
[pip3] flake8-eradicate==1.5.0
[pip3] flake8-isort==7.0.0
[pip3] flake8-logging==1.8.0
[pip3] flake8-noqa==1.4.0
[pip3] flake8-pie==0.16.0
[pip3] flake8-plugin-utils==1.3.3
[pip3] flake8-pytest-style==2.2.0
[pip3] flake8-quotes==3.4.0
[pip3] flake8-return==1.2.0
[pip3] flake8_simplify==0.30.0
[pip3] intel-cmplr-lib-ur==2025.3.1
[pip3] intel-openmp==2025.3.1
[pip3] mkl==2025.3.0
[pip3] mkl-include==2025.3.0
[pip3] mkl-service==2.6.1
[pip3] mkl-static==2025.3.0
[pip3] mypy==1.19.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.1
[pip3] nunchaku==1.2.0+torch2.10
[pip3] nvidia-cuda-runtime==13.1.80
[pip3] onemkl-license==2025.3.0
[pip3] onnx==1.20.1
[pip3] onnx-ir==0.1.14
[pip3] onnxruntime-gpu==1.23.2
[pip3] onnxscript==0.5.7
[pip3] open_clip_torch==3.2.0
[pip3] optree==0.18.0
[pip3] pytorch-lightning==2.6.0
[pip3] sageattention==2.2.0+cu130torch2.9.0andhigher.post4
[pip3] spas_sage_attn==0.1.0+cu130torch2.9.0andhigher.post4
[pip3] tbb==2022.3.0
[pip3] tbb-devel==2022.3.0
[pip3] tcmlib==1.4.1
[pip3] torch==2.10.0+cu130
[pip3] torch-stoi==0.2.3
[pip3] torchaudio==2.10.0+cu130
[pip3] torchdiffeq==0.2.5
[pip3] torchlibrosa==0.1.0
[pip3] torchmetrics==1.8.2
[pip3] torchsde==0.2.6
[pip3] torchvision==0.25.0+cu130
[pip3] triton-windows==3.6.0.post24
[pip3] umf==1.0.2
[conda] Could not collect
</details>

I've also reproduced it on another PC with AMD GPU, Linux, nightly ROCm and torch.

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by pythongiant** [comment]
What’s broken is that torch.compile (TorchDynamo + Inductor) miscompiles dtype-changing view() operations by treating them as safe aliasing views, even though they reinterpret storage at the byte level. which is why compiling an unrelated BF16 path or adding a “no-op” reshape changes the Q6_K result (the compiled kernel ends up reading misaligned or stale bytes, yielding 0.0625 instead of 1.0)


The proper fix in PyTorch is to treat dtype-changing views as unsafe, either force a materializing copy (or alignment/lifetime barrier) when view(dtype) changes element size, or tighten alias analysis so Inductor cannot reuse storage across such reinterpret casts. 

a practical workaround is to avoid view(dtype) under torch.compile: replace it with to(dtype) (copy), or insert clone() / contiguous() right after any dtype reinterpretation, or keep GGUF dequant code out of compiled regions; all of these prevent buffer reuse and restore correctness at the cost of a small copy.

**Comment by pythongiant** [comment]
@chauhang @penguinwu Does this diagnosis and proposed fix sound correct to you? Should `aten.view.dtype` that changes element size be treated as unsafe in Inductor (copy-on-view or a no-alias/lifetime barrier), rather than a normal aliasing view, to prevent cross-graph buffer reuse and miscompilation in cases like GGUF dequant?

**Comment by anijain2305** [comment]
@eellison @v0i0 @mlazos  This seems like inductor issue

I have tried with aot_eager_inductor_decomps backend and the accuracy passes, can you take a look? This is high prio, not just because of incorrect results, but also because it affects ComfyUI which relies on torch.compile to speedup image-video gen models.

**Comment by eellison** [comment]
cc @laithsakka , i tested this only reprod with dynamic shapes. thus the recompile. can you take a look ?

**Comment by Cheng-Logic** [comment]
TorchDynamo generates identical cache keys (CodeID) for factory-wrapped functions (like wrap_compile) because it hashes the __code__ object without hashing the bound closure variables, causing the second compilation (dequantize_blocks_Q6_K) to falsely hit the cache of the first (dequantize_blocks_BF16) and execute with stale static assumptions (e.g. wrong shape/stride constants for view(dtype)).
```Python
// Fix in torch/_dynamo/utils.py (or where CodeID/CacheKey is defined)
class CodeID:
    # ...
    def __hash__(self):
        # Must hash closure cell contents to differentiate instances generated by the same factory
        closure_hash = tuple(hash(c.cell_contents) for c in self.fn.__closure__) if self.fn.__closure__ else ()
        return hash((self.code_hash, closure_hash))
    
    def __eq__(self, other):
        if not isinstance(other, CodeID): return False
        self_closure = tuple(c.cell_contents for c in self.fn.__closure__) if self.fn.__closure__ else ()
        other_closure = tuple(c.cell_contents for c in other.fn.__closure__) if other.fn.__closure__ else ()
        return self.code_hash == other.code_hash and self_closure == other_closure
```
